api/traces/v1/http.go

```diff
@@ -0,0 +1,291 @@
+		r.Use(c.readMiddlewares...)
+		r.Handle("/*", c.instrument.NewHandler(
+			prometheus.Labels{"group": "tracesv1ui", "handler": "search"},
+			proxyRead))
+	})
+
+	return r
+}
+
+// NewUIHandler creates a trace handler for Jaeger UI that is able to re-write
+// HTML to change the <base> attribute so that it works with the Observatorium-style
+// "/api/v1/traces/{tenant}/" URLs.
+func NewUIHandler(read *url.URL, opts ...HandlerOption) http.Handler {
+	c := &handlerConfiguration{
+		logger:     log.NewNopLogger(),
+		registry:   prometheus.NewRegistry(),
+		instrument: nopInstrumentHandler{},
+	}
+
+	for _, o := range opts {
+		o(c)
+	}
+
+	r := chi.NewRouter()
+
+	var proxyRead http.Handler
+	{
+		level.Debug(c.logger).Log("msg", "Configuring upstream Jaeger UI", "ui", read)
+		middlewares := proxy.Middlewares(
+			proxy.MiddlewareSetUpstream(read),
+			proxy.MiddlewareSetPrefixHeader(),
+			proxy.MiddlewareLogger(c.logger),
+			proxy.MiddlewareMetrics(c.registry, prometheus.Labels{"proxy": "tracesv1-ui"}),
+		)
+
+		t := &http.Transport{
+			DialContext: (&net.Dialer{
+				Timeout: ReadTimeout,
+			}).DialContext,
+		}
+
+		proxyRead = &httputil.ReverseProxy{
+			Director:  middlewares,
+			ErrorLog:  proxy.Logger(c.logger),
+			Transport: otelhttp.NewTransport(t),
+
+			// This is the key piece, it changes <base href=> tags
+			ModifyResponse: jaegerUIResponseModifier,
+		}
+	}
+
+	r.Group(func(r chi.Router) {
+		r.Use(c.readMiddlewares...)
+		r.Handle("/*", c.instrument.NewHandler(
+			prometheus.Labels{"group": "tracesv1ui", "handler": "search"},
+			proxyRead))
+	})
+
+	return r
+}
+
+func jaegerUIResponseModifier(response *http.Response) error {
+	// Only modify successful HTTP
+	if response.StatusCode == http.StatusOK && strings.HasPrefix(response.Header.Get("Content-Type"), "text/html") {
+		// Do man-in-the-middle rewriting of the UI HTML.
+		var err error
+
+		// Uncompressed reader
+		var reader io.ReadCloser
+
+		// Read what Jaeger UI sent back (which might be compressed)
+		switch response.Header.Get("Content-Encoding") {
+		case "gzip":
+			reader, err = gzip.NewReader(response.Body)
+			if err != nil {
+				return err
+			}
+			defer reader.Close()
+		case "deflate":
+			reader = flate.NewReader(response.Body)
+			defer reader.Close()
+		default:
+			reader = response.Body
+		}
+
+		b, err := ioutil.ReadAll(reader)
+		if err != nil {
+			return err
+		}
+
+		// At this point we have read the body.  Even if it didn't have a <body href=>
+		// to modify, we need to create a new Reader.  This code thus executes all the time.
+
+		// JaegerUI insists on a <base>, so create one but use Observatorium's
+		// opinion of the base href, not Jaeger Query's opinion.
+		forwardedPrefix := response.Request.Header.Get(proxy.PrefixHeader)
+
+		// The <base href=> tag generated by Jaeger to tell the UI where to fetch static
+		// assets and query /api
+		const expectedBaseTag = `<base href="/" data-inject-target="BASE_URL"/>`
```

Comment:
is Jaeger used by Obs API in a specific version? or would we expect the `<base href=..>` tag to be changed at some point?

Reply:
The Jaeger API is versioned, but the UI is not.  This handler is for the UI.

I want to bring @pavolloffay into this discussion.  Jaeger does something weird, which was introduced in https://github.com/jaegertracing/jaeger-ui/issues/42 .  The introduced `<base>` to support reverse proxies.  I don't like their solution and I am doing something of the opposite.

To support some reverse proxy, Jaeger introduced a mechanism where the path known to the proxy is told to Jaeger.  So if I have two tenants, `tenant1` and `tenant2`, I can start a Jaeger instance that is told the base URL is `/api/traces/v1/tenant1` and another that is told the base URL is `/api/traces/v1/tenant2`.  The reverse proxy is responsible for remembering the host (or IP) of each tenant and forwarding properly.

What Observatorium does with this PR is offer the endpoint {host}/api/traces/v1/{tenant}/{suffix} and proxy adding header `x-tenant: ...` to {jaeger}/{suffix}.  All queries are sent to the same `--traces.read.endpoint={host:port}`

The approach I am taking will be great when Jaeger has a multi-tenant back-end.  This is in-plan, see https://github.com/jaegertracing/jaeger/issues/3427 .  (We might need a temporary work-around to get to that approach, and I plan to start on that work next).

The other approach is that when we add multitenancy to Jaeger we have Jaeger itself alter the BASE_URL instead of doing it in `jaegerUIResponseModifier()`.  If that approach is taken we still don't need to version the `<base>` tag, we merely delete jaegerUIResponseModifier() from the Observatorium/API code.

@pavolloffay , Does the Jaeger community want to pass the tenant to the UI as a segment in the path or in a header?

Reply:
so we don't have control if this tag gets updated in Jaeger UI right? I think then the response won't get properly modified?

Reply:
> @pavolloffay , Does the Jaeger community want to pass the tenant to the UI as a segment in the path or in a header?

There was no decision on this so far. But I think it will be a header to avoid changing the current APIs.


> What Observatorium does with this PR is offer the endpoint {host}/api/traces/v1/{tenant}/{suffix} and proxy adding header x-tenant: ... to {jaeger}/{suffix}. All queries are sent to the same --traces.read.endpoint={host:port}

As you mentioned this is the final goal but it does not work with the Jaeger at the moment as there is no (soft) multitenancy support in Jaeger. As a temporary solution the API service could use tenant name/id as part (prefix) of the endpoint for the read endpoint.